### PR TITLE
ISPN-14406 Synchronize the partial reindex with refresh and flush

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-14406

### CODE REVIEW

**Comment by fax4ever** [comment]
Thanks @tristantarrant pushed the fix. Let me know if there is anything else we can improve here.

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java [hunk 1]
```diff
@@ -77,18 +81,25 @@ public Void apply(EmbeddedCacheManager embeddedCacheManager) {
          postIndex(indexUpdater, progressState, notifier);
       } else {
          DataConversion keyDataConversion = cache.getKeyDataConversion();
-         Set<Class<?>> classSet = new HashSet<>();
+         Set<Class<?>> classSet = new HashSet<>(keys.size());
+         List<CompletableFuture<Object>> updates = new ArrayList<>(keys.size());
```
Comment (wburns) [inline query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java:85]: Sorry to ping pong back and forth. The preferred way of doing this is with https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/util/concurrent/CompletionStages.java#L44

This way you don't need to create another collection and the stages are as minimal as possible.

**Comment by fax4ever** [comment]
Thank for the idea @wburns. Tried to applied the change here. Let me know if we can do better...

**Comment by tristantarrant** [comment]
Merged, thanks

**Comment by tristantarrant** [comment]
can you backport to 14.0.x please _

**Comment by fax4ever** [comment]
> can you backport to 14.0.x please _

sure!

**Comment by fax4ever** [comment]
> can you backport to 14.0.x please _

https://github.com/infinispan/infinispan/pull/10542

**Pull request merged**